doc/switch_commitment.md

```diff
@@ -0,0 +1,211 @@
+the size of the outputs by 32 bytes.
+
+Later on the mimblewimble mailinglist Tim Ruffing came up with a really 
+[beautiful idea](https://lists.launchpad.net/mimblewimble/msg00479.html) 
+(initially suggested by Pieter Wuille), which offers the same advantages but doesn't 
+need this extra storage of an additional element per unspent output:
+
+The idea is the following: 
+
+A normal Pedersen commitment looks like this:
+
+    v*H + r*G
+
+(`v` is value of the input/output, `r` is a truly random blinding factor and `H` and `G` are
+two generator points on the elliptic curve).
+
+If we adapt this by having `r` not being random itself, but using another random number `r'`
+and create the Pedersen Commitment:
+
+    v*H + r*G
+
+such that:
+
+    r = r' + hash( v*H + r'*G  ,  r'*J )
+    
+(using an additional third generation point `J` on the curve) then `r` still is perfectly 
+valid as a blinding factor (as it's randomly distributed) but now we see
+that the part within the brackets of the hash function (`v*H + r'*G  ,  r'*J`) is an 
+**ElGamal commitment**.
+
+This neat idea lead to the removal of the switch commitment hash from the outputs in this 
+(and following) [pull requests](https://github.com/mimblewimble/grin/issues/998) as now it
+could be easily included into the Pedersen Commitments.
+
+
+This is how it is currently implemented in grin. Pedersen commitments are 
+used for the Confidential Transaction but instead of choosing the blinding factor `r` 
+only by random, it is calculated by adding the hash of an ElGamal commitment to a random `r'` 
+(see here in [main_impl.h#L267](https://github.com/mimblewimble/secp256k1-zkp/blob/73617d0fcc4f51896cce4f9a1a6977a6958297f8/src/modules/commitment/main_impl.h#L267)).
+ 
+
+Switch commitments in general first have been described in the paper 
+["Switch Commitments: A Safety Switch for Confidential Transactions"](https://eprint.iacr.org/2017/237.pdf)).
+The **"switch"** in the name comes from the fact that you can virtually flip a "switch" in
+the future and simply by changing the validation procedure you can change the strength of
+the bindingness property of your commitments (even working in a backwards compatible way with 
+commitments created today).
+
+
+## Summary:
+
+For now grin uses and verifies Pedersen Commitments, like other cryptocurrencies do as well, 
+with the only difference that the random Blinding factor `r` is choosen by some special means.
+
+If some time in the future the bindningness of Pedersen Commitments might no longer be assumed
+secure then grin could simply come up with a Soft-Fork which adds additional 
+validation rules which then require that also the computation of `r` must be validated 
+(i.e. the ElGamal commitment must also be revealed) at the additional costs that the 
+range proofs for ElGamal are more costly (in size and computation) and that the hidingness 
+property is no longer perfectly hiding, but only computational hiding. But for a crypto currency
```

Comment:
This last point is mistaken I think. And it might be good to clarify in a section above as well. We could change our commitment scheme to be already perfectly binding and computationally hiding. But that would be bad. The argument is that a privacy-oriented chain needs to stay private *forever*. Because one could go back to the history of the chain and de-anonymize all transactions if the commitment scheme was broken.

On the other hand, the commitment scheme only needs to be perfectly binding at the time the scheme is broken. By using switch commitments, in case of disaster (the intent being still that nothing is ever broken), at least people get a choice: they can compromise the privacy on their *last* UTXOs, or just leave the money alone and lose it. There are many cases where a privacy leak is much more dangerous to one's life than some cryptocurrency.

Reply:
Hi Ignotus! Thanks for your feedback! 🙂 

I see, I had a wrong understanding of some points here: In my naive thinking I didn't realise that activating the switch commitment validation wouldn't immediately affect all users, but instead give each single user the privilege of having a choice.

So not the coin (or its developers) dictates which of the two mutually exclusive properties (perfectly bindingness or perfectly hidingness) should be prioritised but each user itself decides (in the unfortunate case if the scheme really gets broken).

I will try to reword the text (especially the conclusion) and try to include these considerations.

Reply:
Hi, I reworked parts of the document and tried to give a better understanding of the reasoning behind in 1c88d73. 
Please have a look, when you have some time and let me know if I shall change something. 🙂